File: sdk/python/examples/hello.py
#!/usr/bin/env python3
"""Minimal LatchGate example — lease, execute, print receipt.

Requires a running gate:
    latchgate up      # start gate + dependencies
"""

import asyncio

from latchgate import (
    LatchGateApprovalRequired,
    LatchGateClient,
    LatchGateDenied,
)


async def main() -> None:
    # UDS transport (default) — matches latchgate up.
    # public_base_url must match the server's DPoP htu expectation.
    async with LatchGateClient(public_base_url="http://localhost:3000") as client:
        # 1. Obtain a DPoP-bound lease.
        #    max_calls caps the session budget — the OPA policy denies
        #    further calls once exhausted. Start conservative; the agent
        #    can request a new lease when the budget runs out.
        await client.connect(agent_id="hello-example", max_calls=100)
        print("lease acquired")

        # 2. Execute a protected action.
        try:
            result = await client.execute(
                "http_fetch",
                {"url": "https://httpbin.org/get"},
            )
        except LatchGateDenied as exc:
            print(f"denied by policy: {exc.reason}")
            return
        except LatchGateApprovalRequired as exc:
            print(f"held for approval: {exc.approval_id}")
            return

        print(f"output:  {result.output}")
        print(f"receipt: {result.receipt_id}")
        print(f"verified: {result.verification.get('is_fully_successful', False)}")


if __name__ == "__main__":
    asyncio.run(main())

File: sdk/typescript/examples/hello.ts
/**
 * Minimal LatchGate example — lease, execute, print receipt.
 *
 * Requires a running gate:
 *   latchgate up
 *
 * Run:
 *   npx tsx examples/hello.ts
 */

import { LatchGateClient, LatchGateDenied, LatchGateApprovalRequired } from "../src/index.js";

async function main() {
  // UDS transport (default) — matches latchgate up.
  // publicBaseUrl must match the server's DPoP htu expectation.
  const client = new LatchGateClient({ publicBaseUrl: "http://localhost:3000" });

  try {
    // 1. Obtain a DPoP-bound lease.
    //    maxCalls caps the session budget — the OPA policy denies
    //    further calls once exhausted. Start conservative; the agent
    //    can request a new lease when the budget runs out.
    await client.connect({ agentId: "hello-example", maxCalls: 100 });
    console.log("lease acquired");

    // 2. Execute a protected action.
    let result;
    try {
      result = await client.execute("http_fetch", {
        url: "https://httpbin.org/get",
      });
    } catch (err) {
      if (err instanceof LatchGateDenied) {
        console.log("denied by policy:", err.reason);
        return;
      }
      if (err instanceof LatchGateApprovalRequired) {
        console.log("held for approval:", err.approvalId);
        return;
      }
      throw err;
    }

    console.log("output: ", result.output);
    console.log("receipt:", result.receiptId);
    console.log("verified:", result.verification?.is_fully_successful ?? false);
  } finally {
    await client.close();
  }
}

main().catch(console.error);
